File: app/telegram.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
import hashlib
import hmac
import json
from urllib.parse import parse_qsl

import httpx

from app.budget import category_budget_status, warning_text
from app.db import Database
from app.parser import parse_transaction_message


def validate_init_data(init_data: str, bot_token: str, max_age_seconds: int = 86400) -> dict:
    pairs = dict(parse_qsl(init_data, keep_blank_values=True))
    received_hash = pairs.pop("hash", None)
    if not received_hash:
        raise ValueError("Missing Telegram hash")

    data_check_string = "\n".join(f"{key}={value}" for key, value in sorted(pairs.items()))
    secret_key = hmac.new(b"WebAppData", bot_token.encode(), hashlib.sha256).digest()
    expected_hash = hmac.new(secret_key, data_check_string.encode(), hashlib.sha256).hexdigest()
    if not hmac.compare_digest(expected_hash, received_hash):
        raise ValueError("Invalid Telegram init data")

    auth_date = int(pairs.get("auth_date", "0"))
    if datetime.utcnow().timestamp() - auth_date > max_age_seconds:
        raise ValueError("Telegram init data is expired")

    if "user" in pairs:
        pairs["user"] = json.loads(pairs["user"])
    return pairs


async def send_message(bot_token: str, chat_id: int, text: str, reply_markup: dict | None = None) -> None:
    if not bot_token:
        return
    payload: dict = {"chat_id": chat_id, "text": text}
    if reply_markup:
        payload["reply_markup"] = reply_markup
    async with httpx.AsyncClient(timeout=10) as client:
        await client.post(f"https://api.telegram.org/bot{bot_token}/sendMessage", json=payload)


async def handle_text_message(
    db: Database,
    bot_token: str,
    app_base_url: str,
    warning_ratio: float,
    message: dict,
) -> str:
    chat_id = int(message["chat"]["id"])
    from_user = message.get("from") or {}
    telegram_user_id = int(from_user["id"])
    text = (message.get("text") or "").strip()

    db.upsert_user(telegram_user_id, from_user.get("first_name"), from_user.get("username"))

    if text.startswith("/start"):
        reply = "Send expenses like: food, debit card, lunch, $4.20"
        await send_message(bot_token, chat_id, reply, _dashboard_keyboard(app_base_url))
        return reply

    if text.startswith("/budget"):
        reply = _handle_budget_command(db, telegram_user_id, text)
        await send_message(bot_token, chat_id, reply, _dashboard_keyboard(app_base_url))
        return reply

    try:
        parsed = parse_transaction_message(text)
    except ValueError as exc:
        reply = f"{exc}\nExample: food, debit card, lunch, $4.20"
        await send_message(bot_token, chat_id, reply, _dashboard_keyboard(app_base_url))
        return reply

    db.add_transaction(
        telegram_user_id=telegram_user_id,
        kind=parsed.kind,
        category=parsed.category,
        account=parsed.account,
        description=parsed.description,
        amount=parsed.amount,
        currency=parsed.currency,
    )
    month = Database.current_month()
    status = category_budget_status(db, telegram_user_id, month, parsed.category)
    warning = warning_text(status, warning_ratio)
    amount = abs(parsed.amount)
    reply = f"Saved {parsed.category}: ${amount:.2f} on {parsed.account}."
    if warning:
        reply = f"{reply}\n{warning}"
    await send_message(bot_token, chat_id, reply, _dashboard_keyboard(app_base_url))
    return reply


def _handle_budget_command(db: Database, telegram_user_id: int, text: str) -> str:
    # /budget food 300 or /budget food $300 2026-06
    parts = text.split()
    if len(parts) < 3:
        return "Use: /budget category amount [YYYY-MM]"
    category = parts[1].lower()
    amount = Decimal(parts[2].replace("$", "").replace(",", ""))
    month = parts[3] if len(parts) >= 4 else Database.current_month()
    db.set_budget(telegram_user_id, category, month, amount)
    return f"Budget set for {category}: ${amount:.2f} in {month}."


def _dashboard_keyboard(app_base_url: str) -> dict:
    return {
        "inline_keyboard": [
            [{"text": "Open dashboard", "web_app": {"url": f"{app_base_url.rstrip('/')}/"}}]
        ]
    }

File: lib/telegram.ts
import crypto from "node:crypto";

export type TelegramUser = {
  id: number;
  first_name?: string;
  username?: string;
};

export function validateTelegramInitData(initData: string, botToken: string, maxAgeSeconds = 86400) {
  const params = new URLSearchParams(initData);
  const receivedHash = params.get("hash");
  if (!receivedHash) throw new Error("Missing Telegram init data hash");
  params.delete("hash");

  const dataCheckString = Array.from(params.entries())
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([key, value]) => `${key}=${value}`)
    .join("\n");

  const secretKey = crypto.createHmac("sha256", "WebAppData").update(botToken).digest();
  const expectedHash = crypto.createHmac("sha256", secretKey).update(dataCheckString).digest("hex");
  if (!crypto.timingSafeEqual(Buffer.from(expectedHash), Buffer.from(receivedHash))) {
    throw new Error("Invalid Telegram init data");
  }

  const authDate = Number(params.get("auth_date") || 0);
  const ageSeconds = Math.floor(Date.now() / 1000) - authDate;
  if (!Number.isSafeInteger(authDate) || authDate <= 0 || ageSeconds < 0 || ageSeconds > maxAgeSeconds) {
    throw new Error("Telegram init data is expired");
  }

  const userValue = params.get("user");
  if (!userValue) throw new Error("Telegram user is missing");
  const user = JSON.parse(userValue) as TelegramUser;
  if (!Number.isSafeInteger(user?.id) || user.id <= 0) throw new Error("Telegram user is invalid");

  return {
    user,
    queryId: params.get("query_id")
  };
}

export async function sendTelegramMessage(chatId: number, text: string, replyMarkup?: unknown) {
  const token = process.env.TELEGRAM_BOT_TOKEN;
  if (!token) {
    console.error("TELEGRAM_BOT_TOKEN is not configured; cannot send Telegram message.");
    throw new Error("TELEGRAM_BOT_TOKEN is not configured");
  }

  const response = await fetch(`https://api.telegram.org/bot${token}/sendMessage`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      chat_id: chatId,
      text,
      reply_markup: replyMarkup
    })
  });
  const result = await response.json().catch(() => null);
  if (!response.ok || !result?.ok) {
    console.error("Telegram sendMessage failed", result);
    throw new Error(result?.description || "Telegram sendMessage failed");
  }
}

export async function answerTelegramCallback(callbackQueryId: string, text?: string) {
  const token = process.env.TELEGRAM_BOT_TOKEN;
  if (!token) throw new Error("TELEGRAM_BOT_TOKEN is not configured");

  const response = await fetch(`https://api.telegram.org/bot${token}/answerCallbackQuery`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      callback_query_id: callbackQueryId,
      text
    })
  });
  const result = await response.json().catch(() => null);
  if (!response.ok || !result?.ok) {
    console.error("Telegram answerCallbackQuery failed", result);
    throw new Error(result?.description || "Telegram answerCallbackQuery failed");
  }
}

export async function editTelegramMessageReplyMarkup(chatId: number, messageId: number, replyMarkup?: unknown) {
  const token = process.env.TELEGRAM_BOT_TOKEN;
  if (!token) throw new Error("TELEGRAM_BOT_TOKEN is not configured");

  const response = await fetch(`https://api.telegram.org/bot${token}/editMessageReplyMarkup`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      chat_id: chatId,
      message_id: messageId,
      reply_markup: replyMarkup
    })
  });
  const result = await response.json().catch(() => null);
  if (!response.ok || !result?.ok) {
    console.error("Telegram editMessageReplyMarkup failed", result);
    throw new Error(result?.description || "Telegram editMessageReplyMarkup failed");
  }
}

export function dashboardKeyboard(extraRows: unknown[][] = []) {
  const baseUrl = process.env.APP_BASE_URL || process.env.VERCEL_PROJECT_PRODUCTION_URL || "";
  if (!baseUrl || baseUrl.includes("your-public-domain.example")) {
    return extraRows.length ? { inline_keyboard: extraRows } : undefined;
  }
  const normalized = baseUrl.startsWith("http") ? baseUrl : `https://${baseUrl}`;
  return {
    inline_keyboard: [[{ text: "Open dashboard", web_app: { url: normalized } }], ...extraRows]
  };
}

export function transactionKeyboard(transactionId: number) {
  return dashboardKeyboard([
    [
      { text: "Edit", callback_data: `tx:edit:${transactionId}` },
      { text: "Undo", callback_data: `tx:undo:${transactionId}` }
    ]
  ]);
}
